lib/SILOptimizer/Mandatory/Differentiation.cpp

```diff
@@ -2206,25 +2194,38 @@ static bool diagnoseUnsatisfiedRequirements(ADContext &context,
 /// tuple. Otherwise, add the value directly to `result`.
 static void collectAllExtractedElements(SILValue val,
                                         SmallVectorImpl<SILValue> &result) {
-  if (auto tupleType = val->getType().getAs<TupleType>()) {
-    result.resize(tupleType->getNumElements(), SILValue());
-    for (auto *use : val->getUses())
-      if (auto *tupleExtract = dyn_cast<TupleExtractInst>(use->getUser()))
-        result[tupleExtract->getFieldNo()] = tupleExtract;
-  }
-  else
+  auto tupleType = val->getType().getAs<TupleType>();
+  if (!tupleType) {
     result.push_back(val);
+    return;
+  }
+  result.reserve(tupleType->getNumElements());
+  bool visitedDTI = false;
+  for (auto *use : val->getUses()) {
+    if (auto *dti = dyn_cast<DestructureTupleInst>(use->getUser())) {
+      assert(!visitedDTI && "More than one 'destructure_tuple'!?");
+      result.append(dti->getResults().begin(), dti->getResults().end());
+    }
+  }
+  assert(result.size() == tupleType->getNumElements());
 }
 
 /// Given a value, extracts all elements to `result` from this value if it's a
 /// tuple. Otherwise, add this value directly to `result`.
 static void extractAllElements(SILValue val, SILBuilder &builder,
                                SmallVectorImpl<SILValue> &result) {
-  if (auto tupleType = val->getType().getAs<TupleType>())
-    for (auto i : range(tupleType->getNumElements()))
-      result.push_back(builder.createTupleExtract(val.getLoc(), val, i));
-  else
+  if (auto tupleType = val->getType().getAs<TupleType>()) {
```

Comment:
Just FYI: [another version of `extractAllElements` lives in SILGenPoly.cpp](https://github.com/apple/swift/blob/3374bc730452cc76d33acb9158a6d3ca3f787aee/lib/SILGen/SILGenPoly.cpp#L3363). 
It calls `SILBuilder:: emitDestructureValueOperation`.